### PR TITLE
Refactor algod/kmd/indexer initializers

### PR DESCRIPTION
**Pull request by ivanschuetz**
Removed the builder pattern and moved the initialization to the respective services.

This might look less fancy, but more things are checked at compile time, and it's shorter and easier to understand. There was no discernible benefit in using the builder pattern, having the initializers in a separate module (with the implementation "as is"), or bundling the initialization of different versions under the same namespace. Maybe I'm missing something.

The "custom endpoint" builders are now a regular initializer too (https://github.com/manuelmauro/algonaut/issues/109#issuecomment-980532484).

If in the future we use traits (e.g. for the services, or the client), it could make sense to move the initialization to separate modules, but this seems unlikely (soon at least), and shouldn't be much effort to change if/when it happens.

Up to discussion (just a proposal).

@AlterionX feel free to review/comment as well.

### CODE REVIEW

**Comment by manuelmauro** [comment]
Completely agree with implementing this solution. The single concern I would like to highlight, because maybe you have a solution for it, is the fact that  `Algod::new()` function takes two parameters of the same type and this could lead to ordering errors (editors usually help avoiding this kind of errors though). 
This issue is also shared by `new()` functions in transaction builders.

File: src/algod/v2/mod.rs [hunk 1]
```diff
@@ -238,3 +254,54 @@ impl Algod {
...
+    fn test_client_builder_with_invalid_token() {
+        let res = Algod::new(
+            "http://example.com",
+            "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+        );
+        assert!(res.is_err());
+        assert!(res.err().unwrap() == AlgonautError::BadToken);
+    }
+
+    #[test]
```
Comment (manuelmauro) [inline src/algod/v2/mod.rs:301]: Does this test make sense? Is it allowed to have an empty token when connecting, for instance, to AlgoExplorer's node?

**Review by manuelmauro — APPROVED** [review]

**Comment by ivanschuetz** [comment]
> Completely agree with implementing this solution. The single concern I would like to highlight, because maybe you have a solution for it, is the fact that `Algod::new()` function takes two parameters of the same type and this could lead to ordering errors (editors usually help avoiding this kind of errors though).
> This issue is also shared by `new()` functions in transaction builders.

That's a drawback, but I consider it negligible. It's only 2 parameters (unlikely to become more anytime soon, if ever), the order (url, token) is intuitive, and the strings normally look very different. Errors happen early and are obvious - if you pass incorrect connection data, nothing works.

The builder pattern / named parameters are also not infallible against the level of distraction that would lead to a mistake with the constructor here.

In the transaction builders it's a bit more complicated. We have indeed e.g. some instances of consecutive `Address` (like sender and receiver), which are easier to confuse and less obvious to debug (this is a slightly different topic though, as we actually do need the builder pattern there). Feel free to open an issue - or perhaps reopen https://github.com/manuelmauro/algonaut/issues/109 with a more specific description. One idiomatic solution could be to create newtypes for at least some of those fields, like sender and receiver.

File: src/algod/v2/mod.rs (hunk 1 above)
Comment (ivanschuetz) [reply src/algod/v2/mod.rs:301]: In this case you'd use the `with_headers` initializer and don't pass the token header at all https://github.com/manuelmauro/algonaut/blob/8ba4a464787b8cbdb6aba7e9eefcf1e40c3021eb/src/algod/v2/mod.rs#L29

**Pull request merged**